spring cloud网关中使用北极星限流不生效

版本号：
spring-cloud-starter-gateway  2022.0.4 
spring-cloud-starter-tencent-polaris-ratelimit  1.11.9-2022.0.1

spring-cloud-starter-gateway网关中集成北极星限流，匹配限流规则时限流未生效。后台报错信息： java.lang.IllegalArgumentException: Validation failed for header 'internal-callee-activerule' 

限流规则配置
![a](https://gateway.irked.cn/bus/file/data/upload/2023/09/CztLWIsNVx.png)

> 详细的报错日志贴一下，看起来像是监控上报的时候没有正确读取到生效的限流规则。SCT会在这里塞入生效的限流规则名，然后监控上报时可以上报到北极星。 <img alt="image" width="922" src="https://user-images.githubusercontent.com/24446200/266950736-adbc2490-e17d-4860-979c-be2753ad896f.png">

2023-09-11T16:13:20.857+08:00 ERROR 1 --- [or-http-epoll-4] c.t.c.p.r.f.QuotaCheckReactiveFilter     : fail to invoke getQuota, service is irked-gateway

java.lang.IllegalArgumentException: Validation failed for header 'internal-callee-activerule'
        at io.netty.handler.codec.DefaultHeaders.validateValue(DefaultHeaders.java:1019) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.DefaultHeaders.add(DefaultHeaders.java:330) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.DefaultHeaders.addObject(DefaultHeaders.java:364) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.http.DefaultHttpHeaders.add(DefaultHttpHeaders.java:115) ~[netty-codec-http-4.1.97.Final.jar!/:4.1.97.Final]
        at org.springframework.http.server.reactive.NettyHeadersAdapter.add(NettyHeadersAdapter.java:59) ~[spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.http.server.reactive.NettyHeadersAdapter.add(NettyHeadersAdapter.java:40) ~[spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.http.HttpHeaders.add(HttpHeaders.java:1725) ~[spring-web-6.0.11.jar!/:6.0.11]
        at com.tencent.cloud.polaris.ratelimit.filter.QuotaCheckReactiveFilter.filter(QuotaCheckReactiveFilter.java:122) ~[spring-cloud-starter-tencent-polaris-ratelimit-1.11.9-2022.0.1.jar!/:1.11.9-2022.0.1]
        at org.springframework.web.server.handler.DefaultWebFilterChain.invokeFilter(DefaultWebFilterChain.java:114) ~[spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.server.handler.DefaultWebFilterChain.lambda$filter$0(DefaultWebFilterChain.java:108) ~[spring-web-6.0.11.jar!/:6.0.11]
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:45) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.Mono.subscribe(Mono.java:4495) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.core.publisher.MonoDeferContextual.subscribe(MonoDeferContextual.java:55) ~[reactor-core-3.5.9.jar!/:3.5.9]
        at reactor.netty.http.server.HttpServer$HttpServerHandle.onStateChange(HttpServer.java:1052) ~[reactor-netty-http-1.1.10.jar!/:1.1.10]
        at reactor.netty.ReactorNetty$CompositeConnectionObserver.onStateChange(ReactorNetty.java:710) ~[reactor-netty-core-1.1.10.jar!/:1.1.10]
        at reactor.netty.transport.ServerTransport$ChildObserver.onStateChange(ServerTransport.java:481) ~[reactor-netty-core-1.1.10.jar!/:1.1.10]
        at reactor.netty.http.server.HttpServerOperations.onInboundNext(HttpServerOperations.java:647) ~[reactor-netty-http-1.1.10.jar!/:1.1.10]
        at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:114) ~[reactor-netty-core-1.1.10.jar!/:1.1.10]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at reactor.netty.http.server.HttpTrafficHandler.channelRead(HttpTrafficHandler.java:238) ~[reactor-netty-http-1.1.10.jar!/:1.1.10]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800) ~[netty-transport-classes-epoll-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:509) ~[netty-transport-classes-epoll-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407) ~[netty-transport-classes-epoll-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.97.Final.jar!/:4.1.97.Final]
        at java.base/java.lang.Thread.run(Thread.java:831) ~[na:na]
Caused by: java.lang.IllegalArgumentException: a header value contains prohibited character 0x6d4b at index 0.
        at io.netty.handler.codec.http.DefaultHttpHeaders$HeaderValueValidator.validate(DefaultHttpHeaders.java:395) ~[netty-codec-http-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.http.DefaultHttpHeaders$HeaderValueValidator.validate(DefaultHttpHeaders.java:387) ~[netty-codec-http-4.1.97.Final.jar!/:4.1.97.Final]
        at io.netty.handler.codec.DefaultHeaders.validateValue(DefaultHeaders.java:1017) ~[netty-codec-4.1.97.Final.jar!/:4.1.97.Final]
        ... 47 common frames omitted

2023-09-11T16:13:32.056+08:00 ERROR 1 --- [or-http-epoll-1] m.c.DecodeTransferMetadataReactiveFilter : handle metadata[MetadataContext{fragmentContexts={"transitive":{},"disposable":{"source_service_namespace":"irked-cloud-polaris","source_service_name":"irked-gateway"}}}] error.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,3 +13,4 @@
 - [feat:add swagger exposure filters.](https://github.com/Tencent/spring-cloud-tencent/pull/1146)
 - [feat:add swagger report switch.](https://github.com/Tencent/spring-cloud-tencent/pull/1148)
 - [fix:fix retry loadbalancer not working bug.](https://github.com/Tencent/spring-cloud-tencent/pull/1154)
+- [fix:fix header validation when using Chinese char.](https://github.com/Tencent/spring-cloud-tencent/pull/1166)

--- a/spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/core/EncodeTransferMedataFeignInterceptorTest.java
+++ b/spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/core/EncodeTransferMedataFeignInterceptorTest.java
@@ -47,7 +47,7 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = DEFINED_PORT,
 		classes = EncodeTransferMedataFeignInterceptorTest.TestApplication.class,
-		properties = {"server.port=8081", "spring.config.location = classpath:application-test.yml"})
+		properties = {"server.port=48081", "spring.config.location = classpath:application-test.yml"})
 public class EncodeTransferMedataFeignInterceptorTest {
 
 	@Autowired
@@ -74,7 +74,7 @@ public class EncodeTransferMedataFeignInterceptorTest {
 			return MetadataContextHolder.get().getContext(MetadataContext.FRAGMENT_TRANSITIVE, "b");
 		}
 
-		@FeignClient(name = "test-feign", url = "http://localhost:8081")
+		@FeignClient(name = "test-feign", url = "http://localhost:48081")
 		public interface TestFeign {
 
 			@RequestMapping("/test")

--- a/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/ConfigChangeListenerTest.java
+++ b/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/ConfigChangeListenerTest.java
@@ -47,7 +47,7 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
  */
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = DEFINED_PORT, classes = ConfigChangeListenerTest.TestApplication.class,
-		properties = {"server.port=8081", "spring.config.location = classpath:application-test.yml"})
+		properties = {"server.port=48081", "spring.config.location = classpath:application-test.yml"})
 public class ConfigChangeListenerTest {
 
 	private static final CountDownLatch hits = new CountDownLatch(2);

--- a/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/PolarisConfigRefreshOptimizationListenerNotTriggeredTest.java
+++ b/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/PolarisConfigRefreshOptimizationListenerNotTriggeredTest.java
@@ -58,7 +58,7 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = DEFINED_PORT, classes = PolarisConfigRefreshOptimizationListenerNotTriggeredTest.TestApplication.class,
 		properties = {
-				"server.port=8081",
+				"server.port=48081",
 				"spring.cloud.polaris.address=grpc://127.0.0.1:10081",
 				"spring.cloud.polaris.config.connect-remote-server=false",
 				"spring.cloud.polaris.config.refresh-type=reflect",

--- a/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/PolarisConfigRefreshOptimizationListenerTriggeredTest.java
+++ b/spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/listener/PolarisConfigRefreshOptimizationListenerTriggeredTest.java
@@ -59,7 +59,7 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = DEFINED_PORT, classes = PolarisConfigRefreshOptimizationListenerTriggeredTest.TestApplication.class,
 		properties = {
-				"server.port=8081",
+				"server.port=48081",
 				"spring.cloud.polaris.address=grpc://127.0.0.1:10081",
 				"spring.cloud.polaris.config.connect-remote-server=false",
 				"spring.cloud.polaris.config.refresh-type=reflect",

--- a/spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/filter/QuotaCheckReactiveFilter.java
+++ b/spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/filter/QuotaCheckReactiveFilter.java
@@ -18,6 +18,8 @@
 
 package com.tencent.cloud.polaris.ratelimit.filter;
 
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
 import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.Objects;
@@ -51,6 +53,8 @@ import org.springframework.web.server.ServerWebExchange;
 import org.springframework.web.server.WebFilter;
 import org.springframework.web.server.WebFilterChain;
 
+import static com.tencent.cloud.common.constant.ContextConstant.UTF_8;
+
 /**
  * Reactive filter to check quota.
  *
@@ -58,7 +62,7 @@ import org.springframework.web.server.WebFilterChain;
  */
 public class QuotaCheckReactiveFilter implements WebFilter, Ordered {
 
-	private static final Logger LOGGER = LoggerFactory.getLogger(QuotaCheckReactiveFilter.class);
+	private static final Logger LOG = LoggerFactory.getLogger(QuotaCheckReactiveFilter.class);
 
 	private final LimitAPI limitAPI;
 
@@ -122,22 +126,28 @@ public class QuotaCheckReactiveFilter implements WebFilter, Ordered {
 				response.getHeaders()
 						.add(HeaderConstant.INTERNAL_CALLEE_RET_STATUS, RetStatus.RetFlowControl.getDesc());
 				if (Objects.nonNull(quotaResponse.getActiveRule())) {
-					response.getHeaders()
-							.add(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, quotaResponse.getActiveRule().getName()
-									.getValue());
+					try {
+						String encodedActiveRuleName = URLEncoder.encode(
+								quotaResponse.getActiveRule().getName().getValue(), UTF_8);
+						response.getHeaders().add(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, encodedActiveRuleName);
+					}
+					catch (UnsupportedEncodingException e) {
+						LOG.error("Cannot encode {} for header internal-callee-activerule.",
+								quotaResponse.getActiveRule().getName().getValue(), e);
+					}
 				}
 				return response.writeWith(Mono.just(dataBuffer));
 			}
 			// Unirate
 			if (quotaResponse.getCode() == QuotaResultCode.QuotaResultOk && quotaResponse.getWaitMs() > 0) {
-				LOGGER.debug("The request of [{}] will waiting for {}ms.", path, quotaResponse.getWaitMs());
+				LOG.debug("The request of [{}] will waiting for {}ms.", path, quotaResponse.getWaitMs());
 				waitMs = quotaResponse.getWaitMs();
 			}
 		}
 		catch (Throwable t) {
 			// An exception occurs in the rate limiting API call,
 			// which should not affect the call of the business process.
-			LOGGER.error("fail to invoke getQuota, service is " + localService, t);
+			LOG.error("fail to invoke getQuota, service is " + localService, t);
 		}
 
 		if (waitMs > 0) {

--- a/spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/filter/QuotaCheckServletFilter.java
+++ b/spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/filter/QuotaCheckServletFilter.java
@@ -19,6 +19,8 @@
 package com.tencent.cloud.polaris.ratelimit.filter;
 
 import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
 import java.util.Objects;
 import java.util.Set;
 
@@ -49,6 +51,8 @@ import org.springframework.http.MediaType;
 import org.springframework.lang.NonNull;
 import org.springframework.lang.Nullable;
 import org.springframework.web.filter.OncePerRequestFilter;
+
+import static com.tencent.cloud.common.constant.ContextConstant.UTF_8;
 
 /**
  * Servlet filter to check quota.
@@ -115,8 +119,15 @@ public class QuotaCheckServletFilter extends OncePerRequestFilter {
 				}
 				response.addHeader(HeaderConstant.INTERNAL_CALLEE_RET_STATUS, RetStatus.RetFlowControl.getDesc());
 				if (Objects.nonNull(quotaResponse.getActiveRule())) {
-					response.addHeader(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, quotaResponse.getActiveRule().getName()
-							.getValue());
+					try {
+						String encodedActiveRuleName = URLEncoder.encode(
+								quotaResponse.getActiveRule().getName().getValue(), UTF_8);
+						response.addHeader(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, encodedActiveRuleName);
+					}
+					catch (UnsupportedEncodingException e) {
+						LOG.error("Cannot encode {} for header internal-callee-activerule.",
+								quotaResponse.getActiveRule().getName().getValue(), e);
+					}
 				}
 				return;
 			}

--- a/spring-cloud-tencent-plugin-starters/spring-cloud-tencent-gateway-plugin/src/test/java/com/tencent/cloud/plugin/gateway/SCGPluginsAutoConfigurationTest.java
+++ b/spring-cloud-tencent-plugin-starters/spring-cloud-tencent-gateway-plugin/src/test/java/com/tencent/cloud/plugin/gateway/SCGPluginsAutoConfigurationTest.java
@@ -45,7 +45,7 @@ import static org.springframework.boot.test.context.SpringBootTest.WebEnvironmen
  */
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = DEFINED_PORT, classes = SCGPluginsAutoConfigurationTest.TestApplication.class,
-		properties = {"server.port=8081", "spring.config.location = classpath:application-test.yml",
+		properties = {"server.port=48081", "spring.config.location = classpath:application-test.yml",
 				"spring.cloud.tencent.plugin.scg.staining.rule-staining.enabled = true"})
 public class SCGPluginsAutoConfigurationTest {
 

--- a/spring-cloud-tencent-rpc-enhancement/src/main/java/com/tencent/cloud/rpc/enhancement/plugin/PolarisEnhancedPluginUtils.java
+++ b/spring-cloud-tencent-rpc-enhancement/src/main/java/com/tencent/cloud/rpc/enhancement/plugin/PolarisEnhancedPluginUtils.java
@@ -79,6 +79,7 @@ public final class PolarisEnhancedPluginUtils {
 	private static final List<HttpStatus> HTTP_STATUSES = toList(NOT_IMPLEMENTED, BAD_GATEWAY,
 			SERVICE_UNAVAILABLE, GATEWAY_TIMEOUT, HTTP_VERSION_NOT_SUPPORTED, VARIANT_ALSO_NEGOTIATES,
 			INSUFFICIENT_STORAGE, LOOP_DETECTED, BANDWIDTH_LIMIT_EXCEEDED, NOT_EXTENDED, NETWORK_AUTHENTICATION_REQUIRED);
+
 	private PolarisEnhancedPluginUtils() {
 
 	}
@@ -216,7 +217,15 @@ public final class PolarisEnhancedPluginUtils {
 		if (headers != null && headers.containsKey(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME)) {
 			Collection<String> values = headers.get(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME);
 			if (CollectionUtils.isNotEmpty(values)) {
-				return com.tencent.polaris.api.utils.StringUtils.defaultString(new ArrayList<>(values).get(0));
+				String decodedActiveRuleName = "";
+				try {
+					decodedActiveRuleName = URLDecoder.decode(new ArrayList<>(values).get(0), UTF_8);
+				}
+				catch (UnsupportedEncodingException e) {
+					LOG.error("Cannot decode {} from header internal-callee-activerule.",
+							headers.get(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME).get(0), e);
+				}
+				return com.tencent.polaris.api.utils.StringUtils.defaultString(decodedActiveRuleName);
 			}
 		}
 		return "";

--- a/spring-cloud-tencent-rpc-enhancement/src/test/java/com/tencent/cloud/rpc/enhancement/plugin/PolarisEnhancedPluginUtilsTest.java
+++ b/spring-cloud-tencent-rpc-enhancement/src/test/java/com/tencent/cloud/rpc/enhancement/plugin/PolarisEnhancedPluginUtilsTest.java
@@ -17,9 +17,11 @@
 
 package com.tencent.cloud.rpc.enhancement.plugin;
 
+import java.io.UnsupportedEncodingException;
 import java.net.SocketTimeoutException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.net.URLEncoder;
 import java.util.Arrays;
 import java.util.HashMap;
 
@@ -46,6 +48,7 @@ import org.springframework.context.ApplicationContext;
 import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
 
+import static com.tencent.cloud.common.constant.ContextConstant.UTF_8;
 import static com.tencent.polaris.test.common.Consts.NAMESPACE_TEST;
 import static com.tencent.polaris.test.common.Consts.SERVICE_PROVIDER;
 import static org.assertj.core.api.Assertions.assertThat;
@@ -324,13 +327,14 @@ public class PolarisEnhancedPluginUtilsTest {
 	}
 
 	@Test
-	public void testGetActiveRuleNameFromRequest() {
+	public void testGetActiveRuleNameFromRequest() throws UnsupportedEncodingException {
 
 		HttpHeaders headers = new HttpHeaders();
 		String ruleName = PolarisEnhancedPluginUtils.getActiveRuleNameFromRequest(headers);
 		assertThat(ruleName).isEqualTo("");
 
-		headers.set(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, "mock_rule");
+		String encodedRuleName = URLEncoder.encode("mock_rule", UTF_8);
+		headers.set(HeaderConstant.INTERNAL_ACTIVE_RULE_NAME, encodedRuleName);
 		ruleName = PolarisEnhancedPluginUtils.getActiveRuleNameFromRequest(headers);
 		assertThat(ruleName).isEqualTo("mock_rule");
 	}

